spark/src/main/java/org/apache/iceberg/actions/BaseSparkAction.java

```diff
@@ -128,33 +127,56 @@
     return manifestDF.union(otherMetadataFileDF).union(manifestListDF);
   }
 
+  private static boolean isExpectedCatalogLookupException(Exception exception) {
+    return exception.getMessage().contains("AnalysisException") ||
+        exception.getMessage().contains("SparkException") ||
+        exception.getMessage().contains("NoSuchTableException") ||
+        exception.getMessage().contains("CatalogNotFoundException");
+  }
+
+  // Attempt to use Spark3 Catalog resolution if available on the path
+  private static DynMethods.StaticMethod loadCatalogImpl = null;
+
+  private static Dataset<Row> loadCatalogMetadataTable(SparkSession spark, String tableName, MetadataTableType type)
+      throws NoSuchMethodException {
+    if (loadCatalogImpl == null) {
+      loadCatalogImpl = DynMethods.builder("loadCatalogMetadataTable")
+          .hiddenImpl("org.apache.iceberg.spark.Spark3Util",
+              SparkSession.class, String.class, MetadataTableType.class)
+          .buildStaticChecked();
+    }
+    return loadCatalogImpl.invoke(spark, tableName, type);
+  }
+
   protected static Dataset<Row> loadMetadataTable(SparkSession spark, String tableName, String tableLocation,
                                                   MetadataTableType type) {
-    DataFrameReader noCatalogReader = spark.read().format("iceberg");
+    DataFrameReader dataFrameReader = spark.read().format("iceberg");
     if (tableName.contains("/")) {
       // Hadoop Table or Metadata location passed, load without a catalog
-      return noCatalogReader.load(tableName + "#" + type);
+      return dataFrameReader.load(tableName + "#" + type);
     }
-    // Try catalog based name based resolution
     try {
-      return spark.table(tableName + "." + type);
-    } catch (Exception e) {
-      if (!(e instanceof ParseException || e instanceof AnalysisException)) {
-        // Rethrow unexpected exceptions
-        throw e;
+      // Try DSV2 catalog based name based resolution
+      if (!spark.version().startsWith("2")) {
```

Comment:
Can we invert the condition and check if it starts with 3?

Reply:
but what about Spark 4 :) , let me change it to a numeral check > 2

Reply:
actually I guess since we have an explicit spark3 module we can check just for 3 ...